store/tikv/client.go

```diff
@@ -86,13 +89,110 @@ type connArray struct {
 	v     []*grpc.ClientConn
 	// Bind with a background goroutine to process coprocessor streaming timeout.
 	streamTimeout chan *tikvrpc.Lease
+
+	// For batch commands.
+	batchCommandsCh        chan *batchCommandsEntry
+	batchCommandsClients   []*batchCommandsClient
+	tikvTransportLayerLoad uint64
+}
+
+type batchCommandsClient struct {
+	conn                   *grpc.ClientConn
+	client                 tikvpb.Tikv_BatchCommandsClient
+	batched                sync.Map
+	idAlloc                uint64
+	tikvTransportLayerLoad *uint64
+
+	// Indicates the batch client is closed explicitly or not.
+	closed int32
+	// Protect client when re-create the streaming.
+	clientLock sync.Mutex
+}
+
+func (c *batchCommandsClient) isStopped() bool {
+	return atomic.LoadInt32(&c.closed) != 0
+}
+
+func (c *batchCommandsClient) failPendingRequests(err error) {
+	c.batched.Range(func(key, value interface{}) bool {
+		id, _ := key.(uint64)
+		entry, _ := value.(*batchCommandsEntry)
+		entry.err = err
+		close(entry.res)
+		c.batched.Delete(id)
+		return true
+	})
+}
+
+func (c *batchCommandsClient) batchRecvLoop() {
+	defer func() {
+		if r := recover(); r != nil {
+			buf := tidbutil.GetStack()
+			log.Errorf("batchRecvLoop %v %s", r, buf)
+			metrics.PanicCounter.WithLabelValues(metrics.LabelBatchRecvLoop).Inc()
+		}
+	}()
+
+	for {
+		// When `conn.Close()` is called, `client.Recv()` will returns an error.
+		resp, err := c.client.Recv()
+		if err != nil {
+			if c.isStopped() {
+				return
+			}
+			log.Errorf("batchRecvLoop error when receive: %v", err)
+
+			// Hold the lock to forbid batchSendLoop useing the old client.
+			c.clientLock.Lock()
+			c.failPendingRequests(err) // fail all pending requests.
+			for {                      // try to re-create the streaming in the loop.
+				// Re-establish a application layer stream. TCP layer is handled by gRPC.
+				tikvClient := tikvpb.NewTikvClient(c.conn)
+				streamClient, err := tikvClient.BatchCommands(context.TODO())
+				if err == nil {
+					log.Infof("batchRecvLoop re-create streaming success")
+					c.client = streamClient
+					break
+				}
+				log.Errorf("batchRecvLoop re-create streaming fail: %v", err)
+				// TODO: Use a more smart backoff strategy.
+				time.Sleep(time.Second)
+			}
+			c.clientLock.Unlock()
+			continue
+		}
+
+		responses := resp.GetResponses()
+		for i, requestID := range resp.GetRequestIds() {
+			value, ok := c.batched.Load(requestID)
+			if !ok {
+				panic("batchRecvLoop receives a unknown response")
```

Comment:
Just wondering should we ignore this response instead of panic in such a situation:
we have sent request `a` and `b` to TiKV, and there is a call of `failPendingRequests` before the response of `b` returned back.

Reply:
How about log error here ? @hicqu

Reply:
It's an impossible case. There must be a terrible bug if we get here.